### DISCUSSION TITLE
(question) Power BI connector

### DISCUSSION THREAD

**Discussion by harlequin** [trinodb/trino#13730] [Q&A]
(Question)
We want to use Trino directly with Power BI.

I have seen that there are existing commercial products.

Is there any other way to connect to Power BI and be able to perform direct query.

Or is there a Plan to create an open source ODBC Driver?

**Comment by tooptoop4** [comment]
have u tried https://github.com/openlookeng/hetu-odbc-driver ? "this driver can be connected to other versions (such as prestosql and prestodb) based on presto evolution through configuration, so you can alse use it as a presto odbc driver."

**Comment by harlequin** [comment]
thanks for your suggestions

my further investigations brought that the standard ODBC Connector from Power BI doesn't support "Direct Query" in the cloud.


Out of this I'm thinking to have a "thrift" server which implements the "hive2" interface and forward the requests directly to Trino.

So like a gatway on pure Java level.

Another option would be that Trino it self provides a thrift hive2 compatible endpoint ...

**Comment by buremba** [comment]
@harlequin after some research, it looks like ODBC seems to [support Direct Query.](https://docs.microsoft.com/en-us/power-query/odbc) but it needs to be enabled explicitly switching the direct mode in the connection screen. Hive also makes use of [ODBC connector](https://www.microsoft.com/en-us/download/details.aspx?id=40886). Did you start working on The proxy Hive metastore for Trino or would you like to contribute?

**Comment by harlequin** [comment]
Hey,

I have started to get the source from Apache Spark and take only the thrift server component.

I have started to replace the "thrift" functions with JDBC commands to trino.

Authentication and query execution is working, BUT I have problem to forward the Dataset to PowerBI.


Another big problem, from my point, ..... really "hackish" ;)

If I have a more clean code I will distribute for sure

**Comment by buremba** [comment]
Hi @harlequin, I agree with you as it's being too hackish. :)

I spent some time on it and I believe that the most straightforward way would be through the ODBC driver as PowerBI has native integration with ODBC.

1. Hetu seems to a fork of Presto and they have an [ODBC driver](https://github.com/openlookeng/hetu-odbc-driver). 
2. Elasticsearch (has a similar approach to Trino) also has an ODBC driver here: https://github.com/opendistro-for-elasticsearch/sql/tree/main/sql-odbc
3. The commercial solutions such as Starburst, Qubole, and Teradata seem to work with a third part company called [Magnitude](https://www.magnitude.com/drivers/presto-odbc-jdbc) to be able to provide an ODBC driver for their customers but it's not available for the community version of Trino.

Given that the commercial solutions of Trino also offer an ODBC driver rather than making Trino talk like Hive, I believe that's the correct approach. It looks like already solved the issue so if we can somehow fork their ODBC driver or fork the Mariadb's ODBC driver and make it talk to Trino, that would be the ideal way.

What do you think? I would love to collaborate if you're interested. @findepi not sure you're the right person but since you're an active contributor from Starburst, I would love to hear your opinion.

**Comment by harlequin** [comment]
When you are going with the native ODBC implementation you don't have the Direct Query Modus directly available.

At this point you have also to develop a Power BI Connector which handles the ODBC as direct query source.

I have seen that on Microsoft there is a data source called "HD Insight" which is the same like Apache Spark Thrift just rebranded.

This data source supports direct query and also "cloud usage".


The main problem is that our IT is very restrictive in direction PowerBI and Data Gateways (ODBC driver installation, custom connector etc.).

**Comment by buremba** [comment]
Ah, I see. The custom connector option seems to be [straightforward](https://github.com/microsoft/DataConnectors/tree/master/samples/DirectQueryForSQL) but if you have such limitation, that makes sense. Do you have any plans to open-source the bridge?

**Comment by FleischerT** [comment]
Hi,
is there any news in this regard?

**Comment by jbguerraz** [comment]
Came here for that as well :+1: 
Looks like most of databases OSS projects provide their own implementations of ODBC driver. 
AWS produced a full featured (incl. power bi connector) project here https://github.com/opensearch-project/sql-odbc/tree/main which could be a starting point (sadly, looks like, they didn't open-source their Athena ODBC driver...).
Would love to help with it!
